consensus/src/experimental/buffer_manager.rs

```diff
@@ -137,130 +137,104 @@ impl StateManager {
+    /// Set to None if not exist
+    async fn advance_execution_root(&mut self) {
+        let cursor = self.execution_root.clone().or(self.buffer.head.clone());
+        self.execution_root = find_elem(cursor, |item| item.is_ordered());
+        if self.execution_root.is_some() {
+            let ordered_blocks = get_elem(&self.execution_root).get_blocks().clone();
+            self.execution_phase_tx
+                .send(ExecutionRequest { ordered_blocks })
+                .await
+                .expect("Failed to send execution request")
         }
-        self.execution_root = cursor;
     }
 
-    /// check if the items at and after the signing root is already signed
-    /// if yes, move the signing root to the first *unsigned* item.
-    /// if there is no such item, set it to none.
-    fn try_advance_signing_root(&mut self) {
-        let mut cursor = self.signing_root.clone();
-        while cursor.is_some() {
-            {
-                let buffer_item = get_elem(&cursor);
-                if !matches!(
-                    &*buffer_item,
-                    BufferItem::Signed(_) | BufferItem::Aggregated(_)
-                ) {
-                    break;
+    /// Set the signing root to the first not signed item (Executed) and send execution request
+    /// Set to None if not exist
+    async fn advance_signing_root(&mut self) {
+        let cursor = self.signing_root.clone().or(self.buffer.head.clone());
+        self.signing_root = find_elem(cursor, |item| item.is_executed());
+        if self.signing_root.is_some() {
+            let item = get_elem(&self.signing_root);
+            match item.deref() {
+                BufferItem::Executed(executed_item) => {
+                    let commit_ledger_info = LedgerInfo::new(
+                        executed_item.executed_blocks.last().unwrap().block_info(),
+                        executed_item
+                            .ordered_proof
+                            .ledger_info()
+                            .consensus_data_hash(),
+                    );
+                    self.signing_phase_tx
+                        .send(SigningRequest {
+                            ordered_ledger_info: executed_item.ordered_proof.clone(),
+                            commit_ledger_info,
+                        })
+                        .await
+                        .expect("Failed to send signing request");
                 }
+                _ => unreachable!(),
             }
-            cursor = get_next(&cursor);
         }
-        self.signing_root = cursor;
     }
 
     /// Pop the prefix of buffer items until (including) aggregated_item_cursor
-    /// Prepare an aggregated item and send it to the persisting phase
-    fn try_persisting_blocks(&mut self, aggregated_item_cursor: BufferItemRootType) {
-        let mut cursor = self.buffer.head.as_ref().cloned();
+    /// Send persist request.
+    fn advance_head(&mut self, aggregated_cursor: BufferItemRootType) {
```

Comment:
Maybe we simply pass a block id here? We got the element right in the `process_commit_message `, and here we destruct it again just for the id.